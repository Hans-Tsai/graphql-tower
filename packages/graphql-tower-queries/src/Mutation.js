import _ from 'lodash';
import { GraphQLInputObjectType, GraphQLNonNull, GraphQLObjectType } from 'graphql';
import GraphQLField from './GraphQLField';

export default class Mutation extends GraphQLField {
  constructor() {
    super();

    Object.defineProperty(this, 'resolve', {
      enumerable: true,
      set: (resolve) => { this._.resolve = resolve; },
      get: () => {
        const { middleware, resolve } = this.constructor;
        return async (...args) => {
          await middleware.apply(this, args);
          return resolve.apply(this, args);
        };
      },
    });

    this.name = this.constructor.name;
    this.outputFields = {};
    this.inputFields = {};
  }

  set name(value) {
    _.set(this, 'type.name', `${value}Payload`);
    _.set(this, 'args.input.type.ofType.name', `${value}Input`);
    this._.name = value;
  }

  set outputFields(fields) {
    this.type = new GraphQLObjectType({ name: `${this._.name}Payload`, fields });
  }

  set inputFields(fields) {
    const inputType = new GraphQLInputObjectType({ name: `${this._.name}Input`, fields });
    _.set(this, 'args.input.type', new GraphQLNonNull(inputType));
  }
}
