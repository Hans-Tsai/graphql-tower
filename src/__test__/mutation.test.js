import _ from 'lodash';
import { GraphQLID, GraphQLInt } from 'graphql';
import Mutation from '../mutation';

describe('mutation', () => {
  it('Mutation', async () => {
    const MutationNode = class extends Mutation {};
    const mutation = new MutationNode();
    expect(mutation).toMatchSnapshot();
  });

  it('Mutation with name', async () => {
    const MutationNode = class extends Mutation {
      name = 'Node';
    };
    const mutation = new MutationNode();
    expect(mutation).toMatchSnapshot();
    expect(_.get(mutation, 'type.name')).toBe('NodePayload');
    expect(_.get(mutation, 'args.input.type.ofType.name')).toBe('NodeInput');
  });

  it('Mutation with fields', async () => {
    const MutationNode = class extends Mutation {
      name = 'Node';
      inputFields = {
        id: { type: GraphQLID },
        value: { type: GraphQLInt },
      };
      outputFields = {
        id: { type: GraphQLID },
        value: { type: GraphQLInt },
      };
    };
    const mutation = new MutationNode();
    expect(mutation).toMatchSnapshot();
    expect(mutation.type.getFields()).toHaveProperty('id');
    expect(mutation.type.getFields()).toHaveProperty('value');
    expect(mutation.args.input.type.ofType.getFields()).toHaveProperty('id');
    expect(mutation.args.input.type.ofType.getFields()).toHaveProperty('value');
  });

  it('Mutation with thunk fields', async () => {
    const MutationNode = class extends Mutation {
      name = 'Node';
      inputFields = () => ({
        id: { type: GraphQLID },
        value: { type: GraphQLInt },
      });
      outputFields = () => ({
        id: { type: GraphQLID },
        value: { type: GraphQLInt },
      });
    };
    const mutation = new MutationNode();
    expect(mutation).toMatchSnapshot();
    expect(mutation.type.getFields()).toHaveProperty('id');
    expect(mutation.type.getFields()).toHaveProperty('value');
    expect(mutation.args.input.type.ofType.getFields()).toHaveProperty('id');
    expect(mutation.args.input.type.ofType.getFields()).toHaveProperty('value');
  });
});
