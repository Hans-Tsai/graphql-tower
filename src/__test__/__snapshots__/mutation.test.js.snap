// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation Mutation 1`] = `
MutationNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "name": "",
    "resolve": [Function],
  },
  "args": Object {
    "input": Object {
      "type": "Input!",
    },
  },
  "resolve": [Function],
  "type": "Payload",
}
`;

exports[`mutation Mutation with fields 1`] = `
MutationNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "name": "Node",
    "resolve": [Function],
  },
  "args": Object {
    "input": Object {
      "type": "NodeInput!",
    },
  },
  "resolve": [Function],
  "type": "NodePayload",
}
`;

exports[`mutation Mutation with name 1`] = `
MutationNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "name": "Node",
    "resolve": [Function],
  },
  "args": Object {
    "input": Object {
      "type": "NodeInput!",
    },
  },
  "resolve": [Function],
  "type": "NodePayload",
}
`;

exports[`mutation Mutation with thunk fields 1`] = `
MutationNode {
  "_": Object {
    "afterware": Array [],
    "middleware": Array [],
    "name": "Node",
    "resolve": [Function],
  },
  "args": Object {
    "input": Object {
      "type": "NodeInput!",
    },
  },
  "resolve": [Function],
  "type": "NodePayload",
}
`;
